okhttp/src/test/java/okhttp3/CertificatePinnerTest.java

```diff
@@ -189,4 +189,53 @@
     } catch (SSLPeerUnverifiedException expected) {
     }
   }
+
+  @Test public void checkForHostnameWithDoubleAsterisk() throws Exception {
+    CertificatePinner certificatePinner = new CertificatePinner.Builder()
+        .add("**.example.co.uk", certA1Sha256Pin)
+        .build();
+
+    // Should be pinned:
+    try {
+      certificatePinner.check("foo.example.co.uk", certB1.certificate());
+      fail();
+    } catch (SSLPeerUnverifiedException expected) {}
+    try {
+      certificatePinner.check("foo.bar.example.co.uk", certB1.certificate());
+      fail();
+    } catch (SSLPeerUnverifiedException expected) {}
+    try {
+      certificatePinner.check("foo.bar.baz.example.co.uk", certB1.certificate());
+      fail();
+    } catch (SSLPeerUnverifiedException expected) {}
+
+    // Should not be pinned:
+    certificatePinner.check("uk", certB1.certificate());
+    certificatePinner.check("co.uk", certB1.certificate());
+    certificatePinner.check("example.co.uk", certB1.certificate());
```

Comment:
Is it worth discussing this case for **?

Reply:
I thought that if `*.example.com` did not include `example.com`, it would be confusing for `**.example.com` to include `example.com`.

Reply:
If people used this with Ant file patterns, which show up elsewhere.  It specifically means zero or more.

https://ant.apache.org/manual/dirtasks.html

> To make things a bit more flexible, we add one extra feature, which makes it possible to match multiple directory levels. This can be used to match a complete directory tree, or a file anywhere in the directory tree. To do this, "**" must be used as the name of a directory. When ** is used as the name of a directory in the pattern, it matches zero or more directories. For example: /test/** matches all files/directories under /test/, such as /test/x.java, or /test/foo/bar/xyz.html, but not /xyz.xml.

Reply:
I am happy to implement it either way. I guess as long as it is documented it will be fine.

@swankjesse any thoughts on this?

Reply:
I think it’s easy enough if we say `*.example.com` matches exactly one level, and `**.example.com` matches any number.

Reply:
I'm not sure whether you mean I should implement the change (to have `**.example.com` match `example.com`) or not.

Reply:
### Single star: match exactly one.

`*.example.com`
 - Matches `foo.example.com`
 - Does not match `foo.bar.example.com`, or `example.com`.

### Two stars: match zero or more.

`**.example.com`
Matches `example.com`, `foo.example.com`, and `foo.bar.example.com`.

Reply:
Thanks for the clarification.

Reply:
I have implemented this change.